### PR TITLE
Add VarZeroSliceIter, ZeroSliceIter

### LINKED ISSUE #4487: Add explicit ZeroVec and VarZeroVec iterator types
The standard library exports concrete types for iterators. However, our ZeroVec types only return `impl Iterator`. This limits a certain amount of flexibility; for example, you can't put multiple iterators together into a list because the compiler doesn't know that the `impl Iterator`s are going to be the same type.

I also question whether `impl Iterator` is highly efficient in terms of code size and performance. I remember pulling apart some spaghetti iterators into procedural code and getting a boost. Returning a concrete type could make code simpler to compile.

Thoughts? @Manishearth

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes #4487

### CODE REVIEW

File: utils/zerovec/src/varzerovec/components.rs [hunk 1]
```diff
@@ -524,6 +501,80 @@ impl<'a, T: VarULE + ?Sized, F: VarZeroVecFormat> VarZeroVecComponents<'a, T, F>
...
+            // which means we can treat this data as a T
+            T::from_bytes_unchecked(self.components.things.get_unchecked(self.start_index..end))
+        };
+        self.index += 1;
+        // Invariant upheld: end has the same invariant as self.start_index
+        self.start_index = end;
+        Some(item)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
```
Comment (sffc) [inline utils/zerovec/src/varzerovec/components.rs:566]: Praise: Thanks for adding `size_hint`; this is an improvement over what we had before, I think

**Comment by Manishearth** [comment]
thought: these could be VarZeroIter, ZeroIter as well

**Comment by sffc** [comment]
well, it's an iterator over specifically a VarZeroSlice, not a VarTupleULE or some other VarULE container.

**Review by sffc — APPROVED** [review]
I think `VarZeroSliceIter` and `ZeroSliceIter` are the right names

**Pull request merged**